### PR TITLE
[flang]Pass to add vscale range attribute

### PR DESCRIPTION
**Pull request by Leporacanthicus**
Add vscale range attirbute for the Scalable Vector Extension (SVE) if provided on the command-line (options in a previous commit)

If no command-line option is provided, if the target-feature of SVE is specified and the architecture is AArch64, it defualts to 128-2048. in other words a vscale-min of 1, vscale-max of 16.

A pass is used to add the atribute to all functions. The vectorizer will use this attribute to generate the SVE instruction to match the range specified. The attribute is harmless if there is no vectorizable operations in the function.

### CODE REVIEW

**Review by jeanPerier — COMMENTED** [review]
Pass plugin-in looks ok to me (please address the formatting issue from the build bot). I only wonder a bit if this should not be done in the codegen/target rewrite pipeline rather than in `createDefaultFIROptimizerPassPipeline` though (unless the attribute matters to MLIR/FIR optimizations like inlining ?)

I do not know about enough LLVM vscale range attribute to review the logic of where to set it and with which values, so best if you get approval from an SVE knowledgeable person here.

File: flang/include/flang/Tools/CrossToolHelpers.h [hunk 1]
```diff
@@ -42,6 +42,8 @@ struct MLIRToLLVMPassPipelineConfig {
   bool LoopVersioning = false; ///< Run the version loop pass.
   llvm::codegenoptions::DebugInfoKind DebugInfo =
       llvm::codegenoptions::NoDebugInfo; ///< Debug info generation.
+  unsigned VScaleMin = 0; ///< SVE vector range minimum.
+  unsigned VScaleMax = 0; ///< SVE vector range minimum.
```
Comment (ktkachov-arm) [inline flang/include/flang/Tools/CrossToolHelpers.h:46]: Comment should say "range maximum"

**Review by tblah — COMMENTED** [review]
Looks good. But please address @jeanPerier's comment. This looks to me like it should be more of a codegen thing as well.

**Review by tblah — APPROVED** [review]

**Review by jeanPerier — APPROVED** [review]
Thanks

**Comment by kkwli** [comment]
This commit breaks  [ppc64le-flang-rhel-clang build](https://lab.llvm.org/buildbot/#/builders/21).
```fortran
$ cat end.f
end

$ flang-new -fc1 -triple aarch64-none-linux-gnu -emit-llvm -o - end.f90
error: unable to create target: 'No available targets are compatible with triple "aarch64-none-linux-gnu"'
```

**Comment by Leporacanthicus** [comment]
> This commit breaks [ppc64le-flang-rhel-clang build](https://lab.llvm.org/buildbot/#/builders/21).
> 
> ```fortran
> $ cat end.f90
> end
> 
> $ flang-new -fc1 -triple aarch64-none-linux-gnu -emit-llvm -o - end.f90
> error: unable to create target: 'No available targets are compatible with triple "aarch64-none-linux-gnu"'
> ```

Fix here:
https://github.com/llvm/llvm-project/pull/68306

**Pull request merged**